### ISSUE TITLE
Storybook-addon-jsx doesn't work with knobs

### ISSUE THREAD

**Issue by ktj** [Kilix/storybook-addon-jsx#11]
I can't get this addon to work with knobs addon.
This
```javascript
storiesOf('NavKnobs', module)
  .add('without title', () =>{
    const name = text('Title', 'Page title');
    return (<Nav logo={logo}>
      <Link to="/">Home</Link>
    </Nav>)
  })
// and
storiesOf('NavKnobs', module)
  .addWithJSX('without title', () =>{
    return (<Nav logo={logo}>
      <Link to="/">Home</Link>
    </Nav>)
  })
```
both work.

But this doesn't
```javascript
storiesOf('NavKnobs', module)
  .addWithJSX('without title', () =>{
    const name = text('Title', 'Page title');
    return (<Nav logo={logo}>
      <Link to="/">Home</Link>
    </Nav>)
  })
```
Knobstore is null (after page reload)

**Comment by ktj** [comment]
If you could pass some arguments in addWithJSX, so that I can use those arguments in my stories to bypass knobs before knobstore is initialized,

**Comment by wcastand** [comment]
**Hello ktj,
First, thanks for the interest in this package and for the issue.**

After some investigations, i probably need to change the way the addon is used to handle this case and other addons like this one. 

Using addWithJSX instead of a decorator doesn't allow me to execute the story with the decorator (limitation of the storybook's api). 

I'll update the package to use decorators instead to make it work with other addons. If you could make an repo with your example and your version of storybook/react, ect, that would be cool for testing.

I'll do the change as soon as possible and i'll keep you posted here.

PS: i edit your comment to make it more readable, i don't know if you know this but in markdown you can make your code way more readable.
You need to use ` ```javascript  /*your code */ ``` ` instead of the simple "`".
It will be way more readable for the reader ;)

👍

**Comment by ndelangen** [comment]
> (limitation of the storybook's api).

What do you need?

**Comment by wcastand** [comment]
Hi,
I can't use the storybook API to call story() with the decorators. We might have found a way to work around that but that's not pretty so we will refacto the plugin to be a decorator like all the plugins outhere even if i'm not a fan of plugin as decorators.

Storybook API doesn't let me access decorators or context because .add return the api and not the client_api.
```javascript

    api.add = (storyName, getStory) => {
      if (this._storyStore.hasStory(kind, storyName)) {
        throw new Error(`Story of "${kind}" named "${storyName}" already exists`);
      }

      // Wrap the getStory function with each decorator. The first
      // decorator will wrap the story function. The second will
      // wrap the first decorator and so on.
      const decorators = [...localDecorators, ...this._globalDecorators];

      const fn = decorators.reduce(
        (decorated, decorator) => context => decorator(() => decorated(context), context),
        getStory,
      );

      // Add the fully decorated getStory function.
      this._storyStore.addStory(kind, storyName, fn);
      return api;
    };

    api.addDecorator = decorator => {
      localDecorators.push(decorator);
      return api;
    };

    return api;
  }
```

**Comment by ktj** [comment]
Hello, thanks for the reply.
https://github.com/ktj/jsx-story-test
Here is an example project
```bash
yarn
yarn storybook
```
Uncomment line 19 in stories/Nav.stories.js to produce error.
```javascript
// const test = text('Test knob', 'Some text');
```
(Needs page reload)

**Comment by alexandrebodin** [comment]
@ktj I made a PR #12  that should work Can you try it with your project ?

**Comment by ktj** [comment]
That PR fixes the problem, thank you!

**Comment by wcastand** [comment]
i merged the PR, so storybook-addon-jsx@3.1.0 should be working for you now.

If everything work for you, i let you close this issue.

Thanks again for the interest in this package and for the help :)
